### PR TITLE
#23532 Job to start and end scheduled Experiments

### PR DESCRIPTION
**Pull request by dsilvam**
Job to both start and end scheduled Experiments to do so.  
Integration test included.

### CODE REVIEW

File: dotCMS/src/main/java/com/dotcms/experiments/business/ExperimentsAPIImpl.java [hunk 1]
```diff
@@ -662,6 +674,29 @@ public boolean isAnyExperimentRunning() throws DotDataException {
                 !APILocator.getExperimentsAPI().getRunningExperiments().isEmpty();
     }
 
+    @Override
+    public void endFinalizedExperiments(final User user) throws DotDataException {
```
Comment (freddyucv) [inline dotCMS/src/main/java/com/dotcms/experiments/business/ExperimentsAPIImpl.java:678]: I think this method should go into the Job

File: dotCMS/src/main/java/com/dotmarketing/init/DotInitScheduler.java [hunk 2]
```diff
@@ -635,6 +638,33 @@ private static void scheduleJob(final JobBuilder jobBuilder) {
 		}
 	}
 
+	private static void addStartEndScheduledExperimentsJob(final Scheduler scheduler) {
```
Comment (freddyucv) [inline dotCMS/src/main/java/com/dotmarketing/init/DotInitScheduler.java:641]: I don't now if we should move this method into another class something like DotExperimentInitScheduler and then use this class here

File: dotCMS/src/integration-test/java/com/dotmarketing/quartz/job/StartEndScheduledExperimentsJobTest.java [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+        // create experiment that will end soon
+        Experiment scheduledToEndExperiment = experimentDataGen
+                .scheduling(Scheduling.builder().endDate(NOW_PLUS_TWO_MINUTES).build())
+                .nextPersisted();
+
+        scheduledToEndExperiment = experimentsAPI.start(scheduledToEndExperiment.id().orElseThrow(), APILocator.systemUser());
+
+        assertEquals(Status.RUNNING, scheduledToEndExperiment.status());
+        // wait some minutes for its end date to be reached
+        Thread.sleep(2*60*1000);
```
[the remark this answers is not in the document]
Comment (dsilvam) [reply dotCMS/src/integration-test/java/com/dotmarketing/quartz/job/StartEndScheduledExperimentsJobTest.java:81]: that would not be a valid scheduling

File: dotCMS/src/main/java/com/dotcms/experiments/business/ExperimentsAPIImpl.java (hunk 1 above)
Comment (dsilvam) [reply dotCMS/src/main/java/com/dotcms/experiments/business/ExperimentsAPIImpl.java:678]: why?

File: dotCMS/src/main/java/com/dotcms/experiments/business/ExperimentsAPIImpl.java [hunk 4]
```diff
@@ -662,6 +674,29 @@ public boolean isAnyExperimentRunning() throws DotDataException {
...
+                .filter((experiment -> experiment.scheduling().orElseThrow().endDate().orElseThrow()
+                        .isBefore(Instant.now()))).collect(Collectors.toList());
+
+        finalizedExperiments.forEach((experiment ->
+                Try.of(()->end(experiment.id().orElseThrow(), user)).getOrElseThrow((e)->
+                        new DotStateException("Unable to end Experiment. Cause:" + e))));
+    }
+
+    @Override
+    public void startScheduledToStartExperiments(final User user) throws DotDataException {
```
[the remark this answers is not in the document]
Comment (dsilvam) [reply dotCMS/src/main/java/com/dotcms/experiments/business/ExperimentsAPIImpl.java:689]: why?

File: dotCMS/src/main/java/com/dotmarketing/init/DotInitScheduler.java (hunk 2 above)
Comment (dsilvam) [reply dotCMS/src/main/java/com/dotmarketing/init/DotInitScheduler.java:641]: This is repeating the pattern done for other jobs

**Review by freddyucv — APPROVED** [review]

File: dotCMS/src/integration-test/java/com/dotmarketing/quartz/job/StartEndScheduledExperimentsJobTest.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (victoralfaro-dotcms) [reply dotCMS/src/integration-test/java/com/dotmarketing/quartz/job/StartEndScheduledExperimentsJobTest.java:81]: This means we will add two more minutes to the ITs overall, right?

File: dotCMS/src/main/java/com/dotmarketing/init/DotInitScheduler.java [hunk 5]
```diff
@@ -635,6 +638,33 @@ private static void scheduleJob(final JobBuilder jobBuilder) {
 		}
 	}
 
+	private static void addStartEndScheduledExperimentsJob(final Scheduler scheduler) {
+		try {
+			final String jobName      = "EndFinalizedExperimentsJob";
+			final String triggerName  = "trigger31";
```
Comment (victoralfaro-dotcms) [inline dotCMS/src/main/java/com/dotmarketing/init/DotInitScheduler.java:644]: AccessTokenRenewJob has the trigger name `trigger31` and group name `group31` reserved.

File: dotCMS/src/integration-test/java/com/dotmarketing/quartz/job/StartEndScheduledExperimentsJobTest.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (dsilvam) [reply dotCMS/src/integration-test/java/com/dotmarketing/quartz/job/StartEndScheduledExperimentsJobTest.java:81]: That's correct @victoralfaro-dotcms

File: dotCMS/src/main/java/com/dotmarketing/init/DotInitScheduler.java (hunk 5 above)
Comment (dsilvam) [reply dotCMS/src/main/java/com/dotmarketing/init/DotInitScheduler.java]: Updated to 32

**Review by victoralfaro-dotcms — APPROVED** [review]

**Pull request merged**